dmn-engine/src/main/scala/org/camunda/dmn/evaluation/RelationEvaluator.scala

```diff
@@ -23,21 +23,23 @@ class RelationEvaluator(
         val columns =
           mapEither[(String, ParsedDecisionLogic), (String, Val)](row.columns, {
             case (column, expr) =>
-              eval(expr, context).right
+              eval(expr, context)
                 .map(r => column -> r)
           })
 
-        columns.right.map(values => ValContext(StaticContext(values.toMap)))
+        columns.map(values => ValContext(StaticContext(values.toMap)))
       }
     )
 
-    rows.right
-      .map(ValList)
-      .map { result =>
+    rows.map(ValList) match {
+      case r @ Right(result) =>
         context.audit(relation, SingleEvaluationResult(result))
-
-        result
-      }
+        r
+      case l @ Left(failure) =>
+        context.audit(relation,
+                      SingleEvaluationResult(ValError(failure.message)))
+        l
+    }
```

Comment:
With the suggested audit function, this can be simplified to:

```
    val result = mapEither(
      relation.rows,
      (row: ParsedRelationRow) => {
        val columns =
          mapEither[(String, ParsedDecisionLogic), (String, Val)](row.columns, {
            case (column, expr) =>
              eval(expr, context)
                .map(r => column -> r)
          })

        columns.map(values => ValContext(StaticContext(values.toMap)))
      }
    ).map(ValList)

    context.audit(result, relation)
    result
```